File: reference/src/telemetry/instrumentation.ts
/**
 * OpenTelemetry instrumentation helpers for AI-SDLC.
 * Uses @opentelemetry/api which is a no-op when no SDK is configured.
 */

import {
  trace,
  metrics,
  SpanStatusCode,
  type Span,
  type Tracer,
  type Meter,
} from '@opentelemetry/api';

const TRACER_NAME = 'ai-sdlc-framework';
const METER_NAME = 'ai-sdlc-framework';

/** Get the AI-SDLC tracer instance. */
export function getTracer(): Tracer {
  return trace.getTracer(TRACER_NAME);
}

/** Get the AI-SDLC meter instance. */
export function getMeter(): Meter {
  return metrics.getMeter(METER_NAME);
}

/**
 * Execute an async function within an OpenTelemetry span.
 * Automatically records errors and sets span status.
 */
export async function withSpan<T>(
  name: string,
  attributes: Record<string, string | number | boolean>,
  fn: (span: Span) => Promise<T>,
): Promise<T> {
  const tracer = getTracer();
  return tracer.startActiveSpan(name, { attributes }, async (span) => {
    try {
      const result = await fn(span);
      span.setStatus({ code: SpanStatusCode.OK });
      return result;
    } catch (error) {
      span.setStatus({
        code: SpanStatusCode.ERROR,
        message: error instanceof Error ? error.message : String(error),
      });
      span.recordException(error instanceof Error ? error : new Error(String(error)));
      throw error;
    } finally {
      span.end();
    }
  });
}

/**
 * Execute a synchronous function within an OpenTelemetry span.
 * Automatically records errors and sets span status.
 */
export function withSpanSync<T>(
  name: string,
  attributes: Record<string, string | number | boolean>,
  fn: (span: Span) => T,
): T {
  const tracer = getTracer();
  return tracer.startActiveSpan(name, { attributes }, (span) => {
    try {
      const result = fn(span);
      span.setStatus({ code: SpanStatusCode.OK });
      return result;
    } catch (error) {
      span.setStatus({
        code: SpanStatusCode.ERROR,
        message: error instanceof Error ? error.message : String(error),
      });
      span.recordException(error instanceof Error ? error : new Error(String(error)));
      throw error;
    } finally {
      span.end();
    }
  });
}

File: sdk-python/src/ai_sdlc/telemetry/instrumentation.py
"""OpenTelemetry instrumentation helpers for AI-SDLC."""

from __future__ import annotations

from contextlib import asynccontextmanager
from typing import TYPE_CHECKING, Any

from opentelemetry import metrics, trace
from opentelemetry.trace import StatusCode

if TYPE_CHECKING:
    from collections.abc import AsyncGenerator

TRACER_NAME = "ai-sdlc-framework"
METER_NAME = "ai-sdlc-framework"


def get_tracer() -> trace.Tracer:
    """Get the AI-SDLC tracer instance."""
    return trace.get_tracer(TRACER_NAME)


def get_meter() -> metrics.Meter:
    """Get the AI-SDLC meter instance."""
    return metrics.get_meter(METER_NAME)


@asynccontextmanager
async def with_span(
    name: str,
    attributes: dict[str, str | int | float | bool] | None = None,
) -> AsyncGenerator[trace.Span, None]:
    """Execute within an OpenTelemetry span (async context manager)."""
    tracer = get_tracer()
    with tracer.start_as_current_span(name, attributes=attributes or {}) as span:
        try:
            yield span
            span.set_status(StatusCode.OK)
        except Exception as exc:
            span.set_status(StatusCode.ERROR, str(exc))
            span.record_exception(exc)
            raise


def with_span_sync(
    name: str,
    attributes: dict[str, Any] | None = None,
) -> Any:
    """Create a synchronous span context manager."""
    tracer = get_tracer()
    return tracer.start_as_current_span(name, attributes=attributes or {})
